Enable debugging on windows

alfgrini75 opened the pull request:
Enable debugging through generating source map and serving the web-page through the same docker image which was used to generate the build. This is necessary to get the source-mapping right

This is an alternative https://github.com/forderud/QtWasm/pull/103 but here we avoid copying files, and it is a bit more clear why it works (we are mapping to the same folders as was used when the .map-file was created)

https://github.com/forderud/QtWasm/pull/36/ has previously tried to do something similar, but I think it is better to do it this way.

It has been a bit confusing when I get source code or not, but after I did this "disable cache", results have been reproducible (when it works or not) 
<img width="929" height="123" alt="image" src="https://github.com/user-attachments/assets/bb5eea0e-e5e4-4db6-8f6c-6a960bfee44f" />

Commit:
Generate debug information in Debug & RelWithDebInfo builds

Commit:
Serve the webpage from the docker image in build_cmake.bat

This is a preparation step to be able to debug the sources.

forderud commented:
How is the web browser able to access the sources for debugging if they're not first copied to the folder served by the web server?

alfgrini75 commented:
> How is the web browser able to access the sources for debugging if they're not first copied to the folder served by the web server?

Docker takes care of it due to "volumes:" in the yaml file

forderud commented:
> > How is the web browser able to access the sources for debugging if they're not first copied to the folder served by the web server?
> 
> Docker takes care of it due to "volumes:" in the yaml file

Please then clean up this PR so that it only contains the `wasm.cmake` changes needed and as little other changes as possible. Also, please extend the PR description with clear document of how to get a working debugging environment up & running.

If you want to introduce `docker compose` then please do so in a _separate_ PR, so that we avoid bundling it with enabling of debugging.

qdel commented:
I also think that docker compose is overkill here. There is no real need for it. We can just build a docker run command which does the same stuff.

alfgrini75 commented:
@forderud I split out https://github.com/forderud/QtWasm/pull/104/ as a first step. 

 @qdel I agree that we can do all in a docker run command without the "docker compose". I changed it already here.

forderud commented on build_cmake.bat:
```diff
@@ -6,11 +6,26 @@ mkdir build
+        -w / forderud/qtwasm:latest ^
+        python3 /WebServer.py 8080
+    echo "helloworld app will be available here: http://localhost:8080/build/app/helloworld.html"
+    echo "Stop it with commands docker stop wasmdebug /  docker rm wasmdebug"
```
This will need to be cleaned up a bit more. Comments:
* Suggest moving the common parts of the `docker run` commands into a reused env.var to reduce duplication and enforce consistency.
* If you want to add more `echo` console logging then please move it to a separate PR so that we avoid bundling it with enabling of debugging.

alfgrini75 replied:
> This will need to be cleaned up a bit more. Comments:
> 
> * Suggest moving the common parts of the `docker run` commands into a reused env.var to reduce duplication and enforce consistency.
> * If you want to add more `echo` console logging then please move it to a separate PR so that we avoid bundling it with enabling of debugging.

I added a third commit. I don't know if that is what you meant. I can squash 2nd and 3rd commit, but left them separate for now since I was not quite sure this is what you meant.

forderud replied:
Thanks Alf! Now it's starting to look like something. Please give me 1-2 days to test this on my end and I'll get back to you.

forderud commented:
@alfgrini75 I've now updated `build_cmake.bat` to avoid having to comment-out the "Custom build settings" commands that will be extended for debug-builds. Would you mind rebasing this PR on top of master?

Please then keep the `Default build settings` part of the if-check unchanged, so that it's still possible to test the image with default settings?

forderud commented on build_cmake.bat:
```diff
@@ -3,14 +3,44 @@ rmdir build /s /q
 
-run_webserver.bat
+    :: Serve website from within docker to get consistent source mapping
+    docker run -p 8080:8080 ^
```
please pass `--rm` as done above to avoid risk of old containers filling up the disk.

forderud commented on build_cmake.bat:
```diff
@@ -3,14 +3,44 @@ rmdir build /s /q
+        -v %LOCAL_BUILD_FOLDER%:%DOCKER_DEBUG_BUILD_FOLDER% ^
+        -v %LOCAL_SOURCE_FOLDER%:%DOCKER_DEBUG_SOURCE_FOLDER% ^
+        -w %DOCKER_DEBUG_SERVE_FOLDER% %DOCKER_IMAGE% ^
+        python3 %DOCKER_WEBSERVER_PY% 8080
```
If you want to run `run_webserver.bat` through docker then I suggest that you also do this in the "%BUILD_MODE%"=="Release" case for consistently. Please switch to docker in the first commit in the PR.

forderud commented on build_cmake.bat:
```diff
@@ -22,4 +26,21 @@ if "%BUILD_TYPE%"=="" (
+docker run --rm -p 8080:8080 ^
+    -v %LOCAL_WEBSERVER_PY%:%DOCKER_WEBSERVER_PY% ^
+    -v %LOCAL_BUILD_FOLDER%:%DOCKER_DEPLOY_BUILD_FOLDER% ^
+    -v %LOCAL_SOURCE_FOLDER%:%DOCKER_DEPLOY_SOURCE_FOLDER% ^
```
Why are you mapping `%cd%\build` to `/build` here, whereas it's mapped to `/project/build` on line 26? The same also applies to  `%cd%\sample` that's mapped to `/source` here and to `/project/source` on line 26.

This inconsistency makes it difficult to understand and maintain the script. Is there a way to map the folders more consistently?

alfgrini75 replied:
> Why are you mapping `%cd%\build` to `/build` here, whereas it's mapped to `/project/build` on line 26? The same also applies to `%cd%\sample` that's mapped to `/source` here and to `/project/source` on line 26.
> 
> This inconsistency makes it difficult to understand and maintain the script. Is there a way to map the folders more consistently?

You are right, I think I was overthinking it. There are 4 "filesystems" which have to be taken into account (and be in "sync"), so it is easy to be confused 1) Local filesystem 2) docker filesystem during build 3) docker filesystem during deploy, 4) filesystem as seen by the webpage (everything under "localhost").  I simplified it and I could still debug.

forderud replied:
Thanks! Now it's starting to look better :-)

forderud commented:
Thanks @alfgrini75! I've now modified this PR a bit, so that it _only_ exposes sources for non-default (debug) builds and not for default release builds. That way, we reduce risk of source-code leakage.

The modified version is submitted in #105 that was just merged.

alfgrini75 commented:
> Thanks @alfgrini75! I've now modified this PR a bit, so that it _only_ exposes sources for non-default (debug) builds and not for default release builds. That way, we reduce risk of source-code leakage.
> 
> The modified version is submitted in #105 that was just merged.

Thanks
